workflow/scripts/add_extra_components.py

```diff
@@ -119,6 +130,103 @@ def attach_storageunits(n, costs, elec_opts, investment_year):
         )
 
 
+def attach_phs_storageunits(n: pypsa.Network, elec_opts):
+    carriers = elec_opts["extendable_carriers"]["StorageUnit"]
+    carriers = [k for k in carriers if "PHS" in k]
+
+    for carrier in carriers:
+        max_hours = int(carrier.split("hr_")[0])
+
+        psh_resources = (
+            gpd.read_file(
+                f"./repo_data/psh/40-100-dam-height-{max_hours}hr-no-croplands-no-ephemeral-no-highways.gpkg"
+            )
+            .to_crs(4326)
+            .rename(
+                columns={
+                    "System Installed Capacity (Megawatts)": "potential_mw",
+                    "System Energy Storage Capacity (Gigawatt hours)": "potential_gwh",
+                    "System Cost (2020 US Dollars per Installed Kilowatt)": "cost_kw",
+                    "Longitude": "longitude",
+                    "Latitude": "latitude",
+                },
+            )
+        )[
+            [
+                "longitude",
+                "latitude",
+                "potential_gwh",
+                "potential_mw",
+                "cost_kw",
+                "geometry",
+            ]
+        ]
+
+        # Round CAPEX to $500 interval
+        psh_resources["cost_kw_round"] = (psh_resources["cost_kw"] / 500).round() * 500
+
+        # Join SC to PyPSA cluster
+        region_onshore = gpd.read_file(snakemake.input.regions_onshore)
+        region_onshore_psh = gpd.sjoin(
+            region_onshore, psh_resources, how="inner"
+        ).reset_index(drop=True)
+        region_onshore_psh_grp = (
+            region_onshore_psh.groupby(["name", "cost_kw_round"])["potential_mw"]
+            .agg("sum")
+            .reset_index()
+        )
+
+        region_onshore_psh_grp["class"] = (
+            region_onshore_psh_grp.groupby(["name"]).cumcount() + 1
+        )
+        region_onshore_psh_grp["class"] = "c" + region_onshore_psh_grp["class"].astype(
+            str
+        )
+        region_onshore_psh_grp["tech"] = carrier
+        region_onshore_psh_grp["carrier"] = region_onshore_psh_grp[
+            ["tech", "class"]
+        ].agg("_".join, axis=1)
+        region_onshore_psh_grp["Generator"] = (
+            region_onshore_psh_grp["name"] + " " + region_onshore_psh_grp["carrier"]
+        )
+        region_onshore_psh_grp = region_onshore_psh_grp.set_index("Generator")
+
+        # Updated annualize capital cost based on real location
```

Comment:
Is there a reason we use this data instead of the ones from ATB? Or ATB doesnt have these values?

Reply:
Thanks for the clarification! I guess Im still not quite understanding, though. [ATB has FOM and VOM for pumped hydro storage](https://atb.nrel.gov/electricity/2023/pumped_storage_hydropower)? Also, I dont get how these are site specific when we hardcode these values in? 

It may just be me being unfamilar with how PSH data is pulled in! In which case, please feel free to just resolve this note! But a comment in the code may be good. I understand `# Updated annualize capital cost based on real location` to mean costs should be pulled from data, rather than hardcoded in?

Reply:
@trevorb1 sorry for the late response (I was head down for another task). So basically, the CAPEX data for PSH has location-specific costs. Therefore, instead of using a single value that applies to all places, I used location-specific costs in each clustered region. However, the data lacks FOM and VOM information, so I opted to use the ATB's data for all locations.

I will add some comments in the code; hopefully, will resolve some confusion!